src/Shared/Components/PrerenderComponentApplicationStore.cs

```diff
@@ -23,33 +30,103 @@ public PrerenderComponentApplicationStore(string existingState)
                 throw new ArgumentNullException(nameof(existingState));
             }
 
-            ExistingState = JsonSerializer.Deserialize<Dictionary<string, byte[]>>(Convert.FromBase64String(existingState)) ??
-                throw new ArgumentException(nameof(existingState));
+            DeserializeState(Convert.FromBase64String(existingState));
+        }
+
+        protected void DeserializeState(byte[] existingState)
+        {
+            var state = JsonSerializer.Deserialize<Dictionary<string, byte[]>>(existingState);
+            if (state == null)
+            {
+                throw new ArgumentException("Could not deserialize state correctly", nameof(existingState));
+            }
+
+            var stateDictionary = new Dictionary<string, ReadOnlySequence<byte>>();
+            foreach (var (key, value) in state)
+            {
+                stateDictionary.Add(key, new ReadOnlySequence<byte>(value));
+            }
+
+            ExistingState = stateDictionary;
         }
 
 #nullable enable
-        public string? PersistedState { get; private set; }
+        public ReadOnlyMemory<char> PersistedState { get; private set; }
 #nullable disable
 
-        public Dictionary<string, byte[]> ExistingState { get; protected set; }
+        public Dictionary<string, ReadOnlySequence<byte>> ExistingState { get; protected set; }
 
-        public Task<IDictionary<string, byte[]>> GetPersistedStateAsync()
+        public Task<IDictionary<string, ReadOnlySequence<byte>>> GetPersistedStateAsync()
         {
-            return Task.FromResult((IDictionary<string, byte[]>)ExistingState);
+            return Task.FromResult((IDictionary<string, ReadOnlySequence<byte>>)ExistingState);
         }
 
-        protected virtual byte[] SerializeState(IReadOnlyDictionary<string, byte[]> state)
+        protected virtual PooledByteBufferWriter SerializeState(IReadOnlyDictionary<string, ReadOnlySequence<byte>> state)
         {
-            return JsonSerializer.SerializeToUtf8Bytes(state);
+            // System.Text.Json doesn't support serializing ReadonlySequence<byte> so we need to buffer
+            // the data with a memory pool here. We will change our serialization strategy in the future here
+            // so that we can avoid this step.
+            var buffer = new PooledByteBufferWriter();
+            var jsonWriter = new Utf8JsonWriter(buffer);
+            jsonWriter.WriteStartObject();
+            foreach (var (key, value) in state)
+            {
+                if (value.IsSingleSegment)
+                {
+                    jsonWriter.WriteBase64String(key, value.First.Span);
+                }
+                else
+                {
+                    WriteMultipleSegments(jsonWriter, key, value);
+                }
+                jsonWriter.Flush();
+            }
```

Comment:
For the logic here generally, if I followed the flow correctly it looks like this is where we JSON-serialize user-supplied data. We should always assume that such JSON-serializing may throw. Is the logic robust enough not to leak any undisposed resources in that case?

Reply:
No, the user data has already been serialized to bytes contained in individual buffers. We have a `IDictionary<string, PooledByteBufferWriter>` and we are serializing all that into a big json blob object to put it in the response.